### PR TITLE
adding pod resources tuning options, adding imagepullsecert to promexporter

### PR DESCRIPTION
**Pull request by sepulworld**
Signed-off-by: Zane Williamson <zanew@zillow.com>

*Issue number of the reported bug or feature request: https://github.com/bloomberg/solr-operator/issues/34

**Describe your changes**
Added ImagePullSecrets and Pod Resource options to Deployment for PrometheusExporter
Added Pod Resource options to StatefulSet generation SolrCloud

**Testing performed**
Tested changes locally and in environments

**Additional context**
Validated on K8s 1.15

### CODE REVIEW

**Comment by sepulworld** [comment]
Perhaps before merging this, let me look into adding Pod resource options that can be passed to zk-operator and etcd-operator as well.

**Comment by HoustonPutman** [comment]
I would probably change the name `pod` to `solrPodPolicy`, since we have been namespacing the solr properties with Solr. And there are other pods in the pipeline, such as the busybox pod, so it makes sense.

But this is definitely something I've been wanting to add myself, nice job!

> let me look into adding Pod resource options that can be passed to zk-operator and etcd-operator as well

Let me know when you've made a decision on this front

**Comment by sepulworld** [comment]
@HoustonPutman we should be good to go with this now. How does it look? Added etcd, zetcd and zookeeper pod policy (resources and affinity) options. Tested out all locally and works as expected! Let me know what you think. If we merge this, I can work with @swarupdonepudi to rebase his PR with my changes here.

File: pkg/apis/solr/v1beta1/solrcloud_types.go [hunk 1]
```diff
@@ -300,6 +318,22 @@ type ZookeeperSpec struct {
 	// PersistentVolumeClaimSpec is the spec to describe PVC for the zk container
 	// This field is optional. If no PVC spec, etcd container will use emptyDir as volume
 	PersistentVolumeClaimSpec *corev1.PersistentVolumeClaimSpec `json:"persistentVolumeClaimSpec,omitempty"`
+
+	// Pod resources for zookeeper pod
```
Comment (HoustonPutman) [inline pkg/apis/solr/v1beta1/solrcloud_types.go]: probably add the  `// +optional` comment

File: pkg/apis/solr/v1beta1/solrcloud_types.go [hunk 2]
```diff
@@ -364,6 +398,22 @@ type EtcdSpec struct {
 	// PersistentVolumeClaimSpec is the spec to describe PVC for the zk container
 	// This field is optional. If no PVC spec, etcd container will use emptyDir as volume
 	PersistentVolumeClaimSpec *corev1.PersistentVolumeClaimSpec `json:"persistentVolumeClaimSpec,omitempty"`
+
+	// Pod resources for etcd pods
+	EtcdPod EtcdPodPolicy `json:"etcdPodPolicy,omitempty"`
```
Comment (HoustonPutman) [inline pkg/apis/solr/v1beta1/solrcloud_types.go]: another `// +optional` comment

File: pkg/apis/solr/v1beta1/solrcloud_types.go [hunk 3]
```diff
@@ -388,6 +438,22 @@ type ZetcdSpec struct {
 
 	// +optional
 	Image *ContainerImage `json:"image,omitempty"`
+
+	// Pod resources for zetcd pods
+	ZetcdPod ZetcdPodPolicy `json:"zetcdPodPolicy,omitempty"`
```
Comment (HoustonPutman) [inline pkg/apis/solr/v1beta1/solrcloud_types.go]: another `// +optional` comment

File: pkg/apis/solr/v1beta1/solrprometheusexporter_types.go [hunk 4]
```diff
@@ -35,6 +35,11 @@ type SolrPrometheusExporterSpec struct {
 	// +optional
 	Image *ContainerImage `json:"image,omitempty"`
 
+	// Pod defines the policy to create pod for the SolrCloud.
+	// Updating the Pod does not take effect on any existing pods.
+	// +optional
+	Pod SolrPodPolicy `json:"pod,omitempty"`
```
Comment (HoustonPutman) [inline pkg/apis/solr/v1beta1/solrprometheusexporter_types.go]: maybe rename to `podPolicy` or something?

**Comment by sepulworld** [comment]
> Looks good, just a few comments.

Thank you! Updates pushed in

**Pull request merged**